pkg/syncer/cnsoperator/controller/cnsregistervolume/cnsregistervolume_controller_test.go

```diff
@@ -788,6 +789,209 @@ var _ = Describe("checkExistingPVCDataSourceRef", func() {
+		})
+	})
+
+	Context("when topologies are compatible", func() {
+		BeforeEach(func() {
+			pvc.Annotations = map[string]string{
+				"csi.vsphere.volume-accessible-topology": `[{"topology.kubernetes.io/zone":"zone-a"}]`,
+			}
+			mockTopologyMgr.returnTopology = []map[string]string{
+				{"topology.kubernetes.io/zone": "zone-a"},
+			}
+		})
+
+		It("should return nil without error", func() {
+			err := validatePVCTopologyCompatibility(ctx, pvc, volumeDatastoreURL, mockTopologyMgr, mockVC)
+			Expect(err).To(BeNil())
+		})
+	})
+
+	Context("when topologies are incompatible", func() {
+		BeforeEach(func() {
+			pvc.Annotations = map[string]string{
+				"csi.vsphere.volume-accessible-topology": `[{"topology.kubernetes.io/zone":"zone-a"}]`,
+			}
+			mockTopologyMgr.returnTopology = []map[string]string{
+				{"topology.kubernetes.io/zone": "zone-b"},
+			}
+		})
+
+		It("should return error for incompatible zones", func() {
+			err := validatePVCTopologyCompatibility(ctx, pvc, volumeDatastoreURL, mockTopologyMgr, mockVC)
+			Expect(err).ToNot(BeNil())
+			Expect(err.Error()).To(ContainSubstring("is not compatible with volume placement"))
+		})
+	})
+})
+
+var _ = Describe("isTopologyCompatible", func() {
+	It("should return true when all volume segments exist in PVC segments", func() {
+		pvcTopologySegments := []map[string]string{
+			{"topology.kubernetes.io/zone": "zone-a"},
+			{"topology.kubernetes.io/zone": "zone-b"},
+		}
+		volumeTopologySegments := []map[string]string{
+			{"topology.kubernetes.io/zone": "zone-a"},
+		}
+		Expect(isTopologyCompatible(pvcTopologySegments, volumeTopologySegments)).To(BeTrue())
+	})
+
+	It("should return true when volume has multiple segments all existing in PVC", func() {
+		pvcTopologySegments := []map[string]string{
+			{"topology.kubernetes.io/zone": "zone-a", "topology.kubernetes.io/region": "us-west"},
+			{"topology.kubernetes.io/zone": "zone-b", "topology.kubernetes.io/region": "us-west"},
+		}
+		volumeTopologySegments := []map[string]string{
+			{"topology.kubernetes.io/zone": "zone-a"},
+			{"topology.kubernetes.io/zone": "zone-b"},
+		}
+		Expect(isTopologyCompatible(pvcTopologySegments, volumeTopologySegments)).To(BeTrue())
+	})
+
+	It("should return false when volume segment does not exist in any PVC segment", func() {
+		pvcTopologySegments := []map[string]string{
+			{"topology.kubernetes.io/zone": "zone-a"},
+		}
+		volumeTopologySegments := []map[string]string{
+			{"topology.kubernetes.io/zone": "zone-b"},
+		}
+		Expect(isTopologyCompatible(pvcTopologySegments, volumeTopologySegments)).To(BeFalse())
+	})
+
+	It("should return false when some volume segments do not exist in PVC segments", func() {
+		pvcTopologySegments := []map[string]string{
+			{"topology.kubernetes.io/zone": "zone-a"},
+		}
+		volumeTopologySegments := []map[string]string{
+			{"topology.kubernetes.io/zone": "zone-a"},
+			{"topology.kubernetes.io/zone": "zone-b"}, // This doesn't exist in PVC
```

Comment:
for non-zonal policy, even if you request volume on zone-a, and zone-a has shared storage accessible to zone-b, and volume gets provisioned on zone-b, then volume may have zone-a and zone-b as affinity rules.

Reply:
In this example, PVC annotation does not tell us zone-a has shared storage accessible to zone-b though.  How can we express that in PVC annotation?  What would be a proper test case?

Reply:
The current logic tries to find a matching segment in PVC annotation for each segment in the existing volume. Does it mean we only need to find one match and exit?

Reply:
Looks like that is not right.